### ISSUE TITLE
adding Layernorm/BatchNorm compatible with Ragged tensor

### ISSUE THREAD

**Issue by thegodone** [aimat-lab/gcnn_keras#5]
Can you implement LayerNorm and. BatchNorm compatible with RaggedTensor

**Comment by PatReis** [comment]
Some keras layers allow already full RaggedTensors, somethin liek e.g. Add() or Concatenate(). However, I found that this comes with a significant performance decrease, which makes sense, since this should work with arbitrary RaggedTensors.

So what I did for now is: in kgcnn.layers.keras equivalent layers are defined which simply check for ragged_rank=1 (like almost all RaggedTensors with shape (batch, None, F) in kgcnn) and operate directly on the RaggedTensor.values with the original Keras Layer as Fallback. So kgcnn.layers.keras.LayerNormalizationshould already work for axis=-1.

I will look into BatchNorm, which may be more tricky.

**Comment by thegodone** [comment]
LayerNorm is already good enough for testing please wait for BN.

**Comment by thegodone** [comment]
LayerNorm does not work after first dense layer.

**Comment by PatReis** [comment]
Really, also not with the latest Verison? Just to be sure it is kgcnn.layers.keras.LayerNormalization. Could you post the error message?

**Comment by thegodone** [comment]
It works but not as expected in term of performance not a bug

Envoyé de mon iPhone

**Comment by PatReis** [comment]
What do you mean exactly? For example for node-features or edge features of shape (batch, (None), Features) you have the expected behaviour, as far as I understand?

```python
from kgcnn.layers.keras import LayerNormalization

test = tf.ragged.constant([[[10.0,10.0,1.0],[2.0,-2.0,0.0]],[[-5.0,15.0,10.0],[15.0,20.0,200.0],[10.0,100.0,1000.0]]],ragged_rank=1)
result = LayerNormalization(axis=-1)(test)
print(result)
```
gives

`<tf.RaggedTensor [[[0.7070871591567993, 0.7070871591567993, -1.4141743183135986], [1.2245153188705444, -1.2245153188705444, 0.0]], [[-1.3728034496307373, 0.980573832988739, 0.39222949743270874], [-0.7359587550163269, -0.6778567433357239, 1.4138154983520508], [-0.8053871989250183, -0.6040403842926025, 1.4094276428222656]]]>`

**Comment by thegodone** [comment]
I was thinking to use it just before GRU part for example. But the
Haste.LayerNormGRU does the trick too.

**Comment by PatReis** [comment]
Okay, I still do not fully understand. Do you mean like speed or generalization power? Is there a special form of LayerNorm in connection with GRU's. If so, I could add this to the layers in kgcnn.

**Comment by thegodone** [comment]
This is a faster and fully optimal cuda code that is in Haste. indeed it's
both for speed and generalization.

Le lun. 28 juin 2021 à 11:13, Patrick Reiser ***@***.***> a
écrit :
